<a href="https://colab.research.google.com/github/Unravel-Pranav/Deep-Learning/blob/main/Single_layer_perceptron.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import tensorflow as tf

# Define variables for weight and bias
w = tf.Variable([0.2], dtype=tf.float32)  # initial weight
b = tf.Variable([-0.2], dtype=tf.float32) # initial bias

# Define the input as a variable or constant, depending on your needs
x = tf.Variable([1.0], dtype=tf.float32)  # or use `tf.constant([1.0], dtype=tf.float32)`

In [9]:
@tf.function
def linear_model(x):
    return w * x + b


In [10]:
print("Predicted output for [1, 2, 3, 4]:", linear_model(tf.constant([1.0, 2.0, 3.0, 4.0])).numpy())

Predicted output for [1, 2, 3, 4]: [0.         0.2        0.40000004 0.6       ]


In [11]:
# Define a function to calculate the model output and loss
@tf.function
def compute_loss(x, y):
    # Linear model
    linear_model = w * x + b
    # Mean Squared Error loss calculation
    squared_deltas = tf.square(linear_model - y)
    loss = tf.reduce_sum(squared_deltas)
    return loss

In [13]:
# Example input and target output
x = tf.constant([1.0, 2.0, 3.0, 4.0], dtype=tf.float32)
y = tf.constant([0.0, -1.0, -2.0, -3.0], dtype=tf.float32)  # desired output

# Compute loss
loss_value = compute_loss(x, y)
print("Loss:", loss_value.numpy())

Loss: 20.16


## Error is 20.16 , Lets update the weights and check error again.

In [16]:
# Define the optimizer
optimizer = tf.optimizers.SGD(learning_rate=0.01)  # 0.01 is the learning rate

In [14]:
# Define the training step function
def train_step():
    with tf.GradientTape() as tape:
        # Compute the linear model and loss
        linear_model = w * x + b
        squared_deltas = tf.square(linear_model - y)
        loss = tf.reduce_sum(squared_deltas)  # MSE loss

    # Compute gradients and update weights
    gradients = tape.gradient(loss, [w, b])
    optimizer.apply_gradients(zip(gradients, [w, b]))
    return loss

In [17]:
# Training loop
for i in range(1000):  # 1000 training epochs
    loss_value = train_step()
    if i % 100 == 0:  # Print the loss every 100 epochs
        print(f"Epoch {i}: Loss = {loss_value.numpy()}")

Epoch 0: Loss = 20.15999984741211
Epoch 100: Loss = 0.1247127428650856
Epoch 200: Loss = 0.011205842718482018
Epoch 300: Loss = 0.0010068798437714577
Epoch 400: Loss = 9.04696062207222e-05
Epoch 500: Loss = 8.12815960671287e-06
Epoch 600: Loss = 7.300958486666786e-07
Epoch 700: Loss = 6.571497124241432e-08
Epoch 800: Loss = 5.90275206491242e-09
Epoch 900: Loss = 5.33532329427544e-10


In [19]:
# Final model parameters
print("Final weight:", w.numpy())
print("Final bias:", b.numpy())
print("Final loss:", loss_value.numpy())

Final weight: [-0.99999714]
Final bias: [0.99999166]
Final loss: 4.7108983e-11


Prediction check

In [20]:
print("Predicted output for [1, 2, 3, 4]:", linear_model(tf.constant([1.0, 2.0, 3.0, 4.0])).numpy())

Predicted output for [1, 2, 3, 4]: [-5.4836273e-06 -1.0000026e+00 -1.9999998e+00 -2.9999969e+00]


Lets check loss this time, As we can see the loss is reduced now.

---



In [21]:
# Example input and target output
x = tf.constant([1.0, 2.0, 3.0, 4.0], dtype=tf.float32)
y = tf.constant([0.0, -1.0, -2.0, -3.0], dtype=tf.float32)  # desired output

# Compute loss
loss_value = compute_loss(x, y)
print("Loss:", loss_value.numpy())

Loss: 4.6611603e-11
